### DISCUSSION TITLE
Is Shadow the right tool for my research question?

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#1426] [Q&A] [answered]
Is Shadow the right tool for my research question?

(FAQ migrated from Shadow documentation.)

**Comment by robgjansen** [comment, marked as answer]
Shadow is a network simulator/emulator hybrid. It runs real applications, but it simulates network and system functions thereby emulating the kernel to the application. The suitability of Shadow to your problem depends upon what exactly you are trying to measure.

Shadow is intended primarily for analyzing changes in application behavior, e.g. application layer queuing, failure modes, or design changes, and how those changes affect the operation of the system and network performance.

If your research relies on, e.g., the accuracy of specific kernel features or kernel parameter settings, or dynamic changes in Internet routing, then Shadow may not be the right choice as it does not precisely model these behaviors. Shadow is also not the best at measuring cryptographic overhead, so if that is desired then it should probably be done more directly as a separate research component.

See also similar questions answered in #764 and #1244, though note that those were answered for the Shadow v1.x architecture, which is a bit different from the Shadow v2.x architecture.

**Reply by lifeoftomi** [reply]
Thanks for this @robgjansen, is it possible to analyze the performance of calls to an application server running inside of a Shadow simulation? The use case would be running an application on each of the nodes in a Shadow simulated network where the nodes can handle HTTP requests from the host machine, communicate with each other and send responses back to the host machine where the Shadow simulation is running.

For each of those HTTP calls, is there a way to analyze (1) the performance of the nodes based on the network's topology and (2) the performance of the application itself when it handles those requests?

**Reply by robgjansen** [reply]
I'm not fully understanding your use case, could you explain more?

Are you trying to connect up some HTTP requests from the real world outside of Shadow, and add those into the Shadow simulation? If so, this is unlikely to give you the expected results, because Shadow runs its own "virtual world" simulation where time moves forward independent of our real world time.

**Reply by lifeoftomi** [reply]
I want to send a request into the application that will run on Shadow's simulated nodes and receive a response.

Then I want to analyze the behaviour/performance of my application while it runs inside Shadow.

**Reply by robgjansen** [reply]
You will need a program or script that will run in Shadow to send the initial request(s) to your nodes. Then, there are various things you can do to profile the performance of your application.
- You can track behavior of your application by e.g. adding log messages to stdout from inside your application. These log messages will be logged in a file (shadow.data/hosts/hostname`) that you can analyze after the simulation.
- If using the most recent version of Shadow from the main branch (recommended), you can also run the linux `perf` tool on your process  to track performance
- You can build shadow using `./setup build --use-perf-timers` to enable some additional course grain timers inside Shadow that Shadow will log as a normal log message.

**Reply by lifeoftomi** [reply]
thanks @robgjansen , I should have mentioned that the request to the nodes will happen multiple times and not just when the network is starting. Then the responses of those requests sent into Shadow also need to be returned to the client outside of Shadow, on the host machine.

**Reply by lifeoftomi** [reply]
Alternatively, is it possible to replace the "static" events Shadow has bootstrapped with my dynamic and frequent requests from outside Shadow?

If that's not possible, then is there a way to "export a simulation's state" such that any new request I have can be treated as the bootstrapped event and then I inject this previously exported state that now contains my new request as a bootstrapped event and then Shadow runs that?

**Reply by robgjansen** [reply]
Shadow is not designed to interact with other processes/events from outside of its internal simulated environment, so what you are asking is "not possible" in the sense that it is not a use-case that we support.

Can you hack something together? Well, yes, many things are possible if you know how to code. :) For example, you could write an application that
- communicates with some external process (in a way that shadow does not hijack, such as `mqueue`)
- consumes some external events from the external process
- redirects those external events into some form of network communication that happens inside of the Shadow environment (using network io that Shadow does hijack, such as TCP sockets)

Honestly, though, Shadow is not designed for this and it seems that what you are trying to do would be easier to do in an emulator. Perhaps you might consider [mininet](http://mininet.org/), which does not restrict you to a simulated environment?

**Reply by lifeoftomi** [reply]
Thanks @robgjansen, I'll check out mininet to hopefully be able to run any application in an interactive simulation. But just to confirm I understand, there's currently no way to export the state of a Shadow simulation, modify it by adding a new event to interact with my application and then inject this modified simulation state back into Shadow to continue running? Which means that Shadow simulations need events to be "hard coded" or scheduled in the configurations before a simulation starts.
